status-im/nimbus
# Gas refactoring - decouple opcode logic and gas

This PR isolates gas computation in a [self-contained file gas_costs](https://github.com/status-im/nimbus/blob/gas-refactoring-decouple-logic-and-gas/nimbus/vm/forks/gas_costs.nim):
  - easier to read and reason about gas
  - declutter opcode logic
  - easier to debug gas related issue

Furthermore apart for the selfDestruct instruction, all gas computation logics are implemented.

## Design choices

All gas handling is stateless:
  - easier to test and fuzz-test gas proc
  - easier to add a gas-evaluation/simulation feature to Nimbus (which seems to be lacking in the Ethereum world)
  - can be multi-threaded (avoid race condition on the computation state)
  - results can be cached/memoized
  - useful if we want to remove inheritance of Ethereum forks (VM and VM_STATE) from the codebase
  - `extendMemory` function that was hiding gas cost for memory expansion has been removed.
    For opcodes like `MStore`, gas cost is now explicitly GasVeryLow + MemoryExpansionCost.

Most gas computation can be done at compile-time:
  - Faster!
  - Less memory allocated at runtime.
  - At the cost of compilation time. (Note that there is an overhaul to add cached incremental compilation in Nim).
  - It should be possible to inline ~all fixed costs, only keep the dynamic gas procs in the final binary and avoid compiling in the gas tables to reduce binary size.~ all gas computations (fixed, dynamic, memExpansion, complex) with a consumeGas template that dispatches to `prefix proc {.inline.}` instead of dispatching through an array of procs (Non-inlinable even though most are very short)

Fork-specific code is easy to add, maintain and has no runtime cost:
  - Derive from a Fee schedule and update only what is needed at compile-time
  - add a `when Fork: fork-specific else: previous-fork-code`
    to create compile-time defined codepath.

Note that now `Op` is a dense enum (with helper macros to reduce boilerplate). This was introduced to workaround with const table bug - https://github.com/nim-lang/Nim/issues/8007.
It has the added benefit of being compatible with [computed gotos](https://nim-lang.org/docs/manual.html#pragmas-computedgoto-pragma), the fastest way to implement an interpreter short of implementing a JIT.
  - See research notes (from Lua, Python, ANTLR, the Bosch x86 interpreter, game emulators, ...): https://github.com/status-im/nimbus/wiki/Interpreter-optimization-resources
  - computedGoto bad codegen with enum with holes: https://github.com/nim-lang/Nim/issues/7699

## Test results

After refactoring, nimbus still passes the default test suite.

Regarding the official JSON test suite

Regressions (several fixed since PR opening):
  - addmod1_overflow2.json
  - ~div1.json~
  - ~fibbonacci_unrolled.json~
  - ~mul7.json~ 
  - mulmod1_overflow.json
  - ~not1.json~

Progressions:
  - signextend_Overflow_dj42.json
  - blockhash257Block.json
  - blockhash258Block.json
  - blockhashInRange.json
  - blockhashMyBlock.json
  - blockhashNotExistingBlock.json
  - blockhashOutOfRange.json
  - coinbase.json

## Upstream bugs & Todos

### Skipped

Note: This PR was supposed to address #47 workaround:
https://github.com/status-im/nimbus/blob/8528f1b70474f4fb77c57bdbd63da56828da2b25/nimbus/opcode.nim#L20-L28

but this will be done separately, the PR is complex enough.

### Bugs uncovered
compile-time object variant are reset at runtime:
  - const table are losing everything https://github.com/nim-lang/Nim/issues/8007
  - const arrays with proc field https://github.com/nim-lang/Nim/issues/8015

### TODO:
  - fix regressions
  - Change `let` to `const` once bug https://github.com/nim-lang/Nim/issues/8015 is fixed https://github.com/status-im/nimbus/blob/b27996ef905553235a4cb0bdab4bfff3ce2a6502/nimbus/vm/forks/gas_costs.nim#L320-L323
    Alternatively dispatch directly to "prefix gasOp" in a consumeGas/refundGas template that avoids the gasCosts array of procs (pointer indirection, harder debug due to anonymous function) and allows inlining gas cost functions. 
  - Remove the special handling of `Call`, `Create`, `SelfDestruct` now that gas has been decoupled.
  - Showcase how to easily add fork-specific code path for gas (with zero runtime cost) on EIP158 hardfork - https://github.com/status-im/nimbus/blob/b27996ef905553235a4cb0bdab4bfff3ce2a6502/nimbus/vm/forks/gas_costs.nim#L276-L288
  - Op are decoded into an Opcode which is a type that only contain a handler now
    Opcode is probably an unnecessary step and we can dispatch directly Op to its handler.
    Assuming we take a similar approach to op handling as what we’ve done to gas handling,
  the op handler will be fork-specific, if so we can inline the gas computation handler at compile-time and be even more efficient. https://github.com/status-im/nimbus/blob/b27996ef905553235a4cb0bdab4bfff3ce2a6502/nimbus/vm_types.nim#L42-L47
  - Instead of using a dispatch table for opcode dispatch, use [Nim computed gotos](https://nim-lang.org/docs/manual.html#pragmas-computedgoto-pragma)
